src/main/java/org/betonquest/betonquest/api/Objective.java

```diff
@@ -257,6 +272,21 @@ public final void addPlayer(final String playerID, final String instructionStrin
      * @param playerID ID of the player
      */
     public final void removePlayer(final String playerID) {
+        removePlayer(playerID, false);
+    }
+
+    /**
+     * Removes the objective from the player. It does not complete it nor update
+     * the database. In order to complete it, use completeObjective() instead.
+     * In order to remove it from database use PlayerData.deleteObjective()
+     * instead.
+     *
+     * @param playerID ID of the player
+     * @param complete is the player remove because he completed the objective
+     */
+    public final void removePlayer(final String playerID, final boolean complete) {
+        final PlayerObjectiveEndEvent event = new PlayerObjectiveEndEvent(PlayerConverter.getPlayer(playerID),this, complete);
```

Comment:
Only issue I see is that there is no possibility to differentiate between a canceled and a paused (opposite to resume) objective in that event.

Reply:
Yes, you are right. Now I have added inner class enum "StartCause" and "EndCause" for the PlayerObjectiveStartEvent and PlayerObjectiveEndEvent. The reasons for the start and end of Objective can now be clearly shown.